File: harness/e2e/zig/recovery_block_harness.zig
const abi = @import("abi");
const checker = @import("checker");
const checkpoint = @import("checkpoint");
const recovery_block = @import("recovery_block");
const std = @import("std");

comptime {
    std.debug.assert(abi.recovery.primary_accepted == @intFromEnum(recovery_block.RecoveryStatus.primary_accepted));
    std.debug.assert(abi.recovery.alternate_accepted == @intFromEnum(recovery_block.RecoveryStatus.alternate_accepted));
    std.debug.assert(abi.recovery.unrecoverable == @intFromEnum(recovery_block.RecoveryStatus.unrecoverable));
    std.debug.assert(abi.recovery.checkpoint_failed == @intFromEnum(recovery_block.RecoveryStatus.checkpoint_failed));
    std.debug.assert(abi.recovery.restore_failed == @intFromEnum(recovery_block.RecoveryStatus.restore_failed));
}

const check = struct {
    const ok: u32 = @intFromEnum(checker.CheckStatus.ok);
    const above_max: u32 = @intFromEnum(checker.CheckStatus.above_max);
    const invalid_checksum: u32 = @intFromEnum(checker.CheckStatus.invalid_checksum);
};

export var harness_iteration: u32 = 0;
export var harness_stage: u32 = 0;
export var harness_fault_target: u32 = 0;
export var harness_fault_value: u32 = 0;
export var harness_last_expected: u32 = 0;
export var harness_last_initial_value: u32 = 0;
export var harness_last_value: u32 = 0;
export var harness_last_status: u32 = 0;
export var harness_last_recovery_status: u32 = 0;
export var harness_last_checkpoint_check: u32 = 0;
export var harness_last_primary_check: u32 = 0;
export var harness_last_restore_check: u32 = 0;
export var harness_last_alternate_check: u32 = 0;
export var harness_last_active_value: u32 = 0;
export var harness_last_checkpoint_value: u32 = 0;
export var harness_passes: u32 = 0;
export var harness_failures: u32 = 0;
export var harness_last_fault_target: u32 = 0;

// Working checkpointed record. Held in a single contiguous struct so it is
// memory-resident at the injection point and corruptible exactly like the C
// harness's `harness_recovery_block_state`. The algorithm runs in place on it;
// mirrorState only publishes the two observable values the tooling reads.
export var harness_recovery_block_state: checkpoint.CheckpointedRecord = undefined;

fn load(ptr: *const volatile u32) u32 {
    return ptr.*;
}

fn store(ptr: *volatile u32, value: u32) void {
    ptr.* = value;
}

fn sampleInitialValue(iteration: u32) u32 {
    return 100 + ((iteration *% 29) % 700);
}

fn sampleRecord(value: u32) checker.CheckedRecord {
    return checker.CheckedRecord.init(
        .sample,
        value,
        0,
        1000,
        6,
        16,
    );
}

fn mirrorState() void {
    store(&harness_last_active_value, harness_recovery_block_state.active.value);
    store(&harness_last_checkpoint_value, harness_recovery_block_state.checkpoint.value);
}

export fn harness_injection_point_before_recovery() callconv(.c) void {
    asm volatile ("nop // injection_point_before_recovery");
}

export fn harness_injection_point_after_recovery() callconv(.c) void {
    asm volatile ("nop // injection_point_after_recovery");
}

fn setPrimaryValueFault(state: *checkpoint.CheckpointedRecord) void {
    state.active.value = load(&harness_fault_value);
}

fn applyAfterPrimaryFault(
    state: *checkpoint.CheckpointedRecord,
    _: *const recovery_block.SampleUpdate,
) void {
    store(&harness_stage, abi.stage.after_primary);

    switch (load(&harness_fault_target)) {
        abi.fault.recovery_primary_value,
        abi.fault.recovery_primary_value_and_alternate_checksum,
        => setPrimaryValueFault(state),
        abi.fault.recovery_primary_checksum => state.active.checksum ^= load(&harness_fault_value),
        abi.fault.recovery_primary_value_and_checkpoint_checksum => {
            setPrimaryValueFault(state);
            state.checkpoint.checksum ^= 0x10;
        },
        else => {},
    }

    mirrorState();
}

fn applyAfterAlternateFault(
    state: *checkpoint.CheckpointedRecord,
    _: *const recovery_block.SampleUpdate,
) void {
    store(&harness_stage, abi.stage.after_alternate);

    if (load(&harness_fault_target) ==
        abi.fault.recovery_primary_value_and_alternate_checksum)
    {
        state.active.checksum ^= 0x10;
    }

    mirrorState();
}

fn incrementPasses() void {
    store(&harness_passes, load(&harness_passes) +% 1);
}

fn incrementFailures() void {
    store(&harness_failures, load(&harness_failures) +% 1);
}

fn validate(initial: u32, expected: u32) void {
    const target = load(&harness_last_fault_target);
    const recovery_status = load(&harness_last_recovery_status);
    const checkpoint_check = load(&harness_last_checkpoint_check);
    const primary_check = load(&harness_last_primary_check);
    const restore_check = load(&harness_last_restore_check);
    const alternate_check = load(&harness_last_alternate_check);
    const active_value = load(&harness_last_active_value);
    const checkpoint_value = load(&harness_last_checkpoint_value);

    switch (target) {
        abi.fault.recovery_primary_value => {
            if (recovery_status == abi.recovery.alternate_accepted and
                checkpoint_check == check.ok and
                primary_check == check.above_max and
                restore_check == check.ok and
                alternate_check == check.ok and
                active_value == expected and
                checkpoint_value == expected)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.recovery_primary_checksum => {
            if (recovery_status == abi.recovery.alternate_accepted and
                checkpoint_check == check.ok and
                primary_check == check.invalid_checksum and
                restore_check == check.ok and
                alternate_check == check.ok and
                active_value == expected and
                checkpoint_value == expected)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.recovery_primary_value_and_alternate_checksum => {
            if (recovery_status == abi.recovery.unrecoverable and
                checkpoint_check == check.ok and
                primary_check == check.above_max and
                restore_check == check.ok and
                alternate_check == check.invalid_checksum and
                active_value == initial and
                checkpoint_value == initial)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.recovery_primary_value_and_checkpoint_checksum => {
            if (recovery_status == abi.recovery.restore_failed and
                checkpoint_check == check.ok and
                primary_check == check.above_max and
                restore_check == check.invalid_checksum and
                alternate_check == check.ok and
                active_value == load(&harness_fault_value))
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        else => {
            if (recovery_status == abi.recovery.primary_accepted and
                checkpoint_check == check.ok and
                primary_check == check.ok and
                restore_check == check.ok and
                alternate_check == check.ok and
                active_value == expected and
                checkpoint_value == expected)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
    }
}

export fn harness_main() callconv(.c) noreturn {
    store(&harness_stage, abi.stage.boot);

    while (true) {
        const iteration = load(&harness_iteration) +% 1;
        const initial = sampleInitialValue(iteration);
        const expected = recovery_block.samplePrimaryValue(iteration);
        var update = recovery_block.SampleUpdate{
            .sample = iteration,
            .faults = recovery_block.sample_fault.none,
        };

        harness_recovery_block_state = checkpoint.CheckpointedRecord.init(sampleRecord(initial));

        store(&harness_iteration, iteration);
        store(&harness_last_initial_value, initial);
        store(&harness_last_expected, expected);
        store(&harness_last_value, 0);
        store(&harness_last_status, abi.recovery.primary_accepted);
        store(&harness_last_recovery_status, abi.recovery.primary_accepted);
        store(&harness_last_checkpoint_check, check.ok);
        store(&harness_last_primary_check, check.ok);
        store(&harness_last_restore_check, check.ok);
        store(&harness_last_alternate_check, check.ok);
        store(&harness_last_fault_target, abi.fault.none);
        mirrorState();

        store(&harness_stage, abi.stage.before_recovery);
        @call(.never_inline, harness_injection_point_before_recovery, .{});

        store(&harness_last_fault_target, load(&harness_fault_target));
        const result = recovery_block.runWithHooks(
            &harness_recovery_block_state,
            &update,
            recovery_block.samplePrimary,
            applyAfterPrimaryFault,
            recovery_block.sampleAlternate,
            applyAfterAlternateFault,
        );

        store(&harness_last_recovery_status, result.status.code());
        store(&harness_last_status, result.status.code());
        store(&harness_last_checkpoint_check, result.checkpoint_check.code());
        store(&harness_last_primary_check, result.primary_check.code());
        store(&harness_last_restore_check, result.restore_check.code());
        store(&harness_last_alternate_check, result.alternate_check.code());
        mirrorState();
        store(&harness_last_value, load(&harness_last_active_value));
        store(&harness_fault_target, abi.fault.none);

        store(&harness_stage, abi.stage.after_recovery);
        validate(initial, expected);
        @call(.never_inline, harness_injection_point_after_recovery, .{});
    }
}

pub fn panic(msg: []const u8, trace: ?*std.builtin.StackTrace, ra: ?usize) noreturn {
    _ = msg;
    _ = trace;
    _ = ra;
    while (true) {}
}

File: harness/e2e/c/recovery_block_harness.c
#include <stdint.h>

#include "harness_abi.h"
#include "../../../c/recovery_block/recovery_block.h"

_Static_assert((int)RECOVERY_BLOCK_PRIMARY_ACCEPTED == HARNESS_RECOVERY_PRIMARY_ACCEPTED,
    "recovery_block_status_t::RECOVERY_BLOCK_PRIMARY_ACCEPTED must match HARNESS_RECOVERY_PRIMARY_ACCEPTED");
_Static_assert((int)RECOVERY_BLOCK_ALTERNATE_ACCEPTED == HARNESS_RECOVERY_ALTERNATE_ACCEPTED,
    "recovery_block_status_t::RECOVERY_BLOCK_ALTERNATE_ACCEPTED must match HARNESS_RECOVERY_ALTERNATE_ACCEPTED");
_Static_assert((int)RECOVERY_BLOCK_UNRECOVERABLE == HARNESS_RECOVERY_UNRECOVERABLE,
    "recovery_block_status_t::RECOVERY_BLOCK_UNRECOVERABLE must match HARNESS_RECOVERY_UNRECOVERABLE");
_Static_assert((int)RECOVERY_BLOCK_CHECKPOINT_FAILED == HARNESS_RECOVERY_CHECKPOINT_FAILED,
    "recovery_block_status_t::RECOVERY_BLOCK_CHECKPOINT_FAILED must match HARNESS_RECOVERY_CHECKPOINT_FAILED");
_Static_assert((int)RECOVERY_BLOCK_RESTORE_FAILED == HARNESS_RECOVERY_RESTORE_FAILED,
    "recovery_block_status_t::RECOVERY_BLOCK_RESTORE_FAILED must match HARNESS_RECOVERY_RESTORE_FAILED");

volatile uint32_t harness_iteration;
volatile uint32_t harness_stage;
volatile uint32_t harness_fault_target;
volatile uint32_t harness_fault_value;
volatile uint32_t harness_last_expected;
volatile uint32_t harness_last_initial_value;
volatile uint32_t harness_last_value;
volatile uint32_t harness_last_status;
volatile uint32_t harness_last_recovery_status;
volatile uint32_t harness_last_checkpoint_check;
volatile uint32_t harness_last_primary_check;
volatile uint32_t harness_last_restore_check;
volatile uint32_t harness_last_alternate_check;
volatile uint32_t harness_last_active_value;
volatile uint32_t harness_last_checkpoint_value;
volatile uint32_t harness_passes;
volatile uint32_t harness_failures;
volatile uint32_t harness_last_fault_target;
volatile checkpoint_record_t harness_c_recovery_block_state;

static uint32_t sample_initial_value(uint32_t iteration) {
    return 100u + ((iteration * 29u) % 700u);
}

static checker_record_t sample_record(uint32_t value) {
    return checker_record_init(
        CHECKER_TAG_SAMPLE,
        value,
        0u,
        1000u,
        6u,
        16u);
}

__attribute__((used, noinline))
void harness_injection_point_before_recovery(void) {
    __asm__ volatile("nop");
}

__attribute__((used, noinline))
void harness_injection_point_after_recovery(void) {
    __asm__ volatile("nop");
}

static void mirror_state(void) {
    const checkpoint_record_t *state =
        (const checkpoint_record_t *)&harness_c_recovery_block_state;
    harness_last_active_value = state->active.value;
    harness_last_checkpoint_value = state->checkpoint.value;
}

static void set_primary_value_fault(checkpoint_record_t *state) {
    state->active.value = harness_fault_value;
}

static void apply_after_primary_fault(checkpoint_record_t *state, void *context) {
    (void)context;

    harness_stage = HARNESS_STAGE_AFTER_PRIMARY;

    switch (harness_fault_target) {
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE:
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE_AND_ALTERNATE_CHECKSUM:
        set_primary_value_fault(state);
        break;
    case HARNESS_FAULT_RECOVERY_PRIMARY_CHECKSUM:
        state->active.checksum ^= harness_fault_value;
        break;
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE_AND_CHECKPOINT_CHECKSUM:
        set_primary_value_fault(state);
        state->checkpoint.checksum ^= 0x10u;
        break;
    default:
        break;
    }

    mirror_state();
}

static void apply_after_alternate_fault(checkpoint_record_t *state, void *context) {
    (void)context;

    harness_stage = HARNESS_STAGE_AFTER_ALTERNATE;

    if (harness_fault_target ==
        HARNESS_FAULT_RECOVERY_PRIMARY_VALUE_AND_ALTERNATE_CHECKSUM) {
        state->active.checksum ^= 0x10u;
    }

    mirror_state();
}

static void increment_passes(void) {
    harness_passes += 1u;
}

static void increment_failures(void) {
    harness_failures += 1u;
}

static void validate(uint32_t initial, uint32_t expected) {
    const uint32_t target = harness_last_fault_target;
    const uint32_t recovery_status = harness_last_recovery_status;
    const uint32_t checkpoint_check = harness_last_checkpoint_check;
    const uint32_t primary_check = harness_last_primary_check;
    const uint32_t restore_check = harness_last_restore_check;
    const uint32_t alternate_check = harness_last_alternate_check;
    const uint32_t active_value = harness_last_active_value;
    const uint32_t checkpoint_value = harness_last_checkpoint_value;

    switch (target) {
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE:
        if (recovery_status == HARNESS_RECOVERY_ALTERNATE_ACCEPTED &&
            checkpoint_check == CHECKER_OK &&
            primary_check == CHECKER_ERR_ABOVE_MAX &&
            restore_check == CHECKER_OK &&
            alternate_check == CHECKER_OK &&
            active_value == expected &&
            checkpoint_value == expected) {
            increment_passes();
        } else {
            increment_failures();
        }
        return;
    case HARNESS_FAULT_RECOVERY_PRIMARY_CHECKSUM:
        if (recovery_status == HARNESS_RECOVERY_ALTERNATE_ACCEPTED &&
            checkpoint_check == CHECKER_OK &&
            primary_check == CHECKER_ERR_INVALID_CHECKSUM &&
            restore_check == CHECKER_OK &&
            alternate_check == CHECKER_OK &&
            active_value == expected &&
            checkpoint_value == expected) {
            increment_passes();
        } else {
            increment_failures();
        }
        return;
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE_AND_ALTERNATE_CHECKSUM:
        if (recovery_status == HARNESS_RECOVERY_UNRECOVERABLE &&
            checkpoint_check == CHECKER_OK &&
            primary_check == CHECKER_ERR_ABOVE_MAX &&
            restore_check == CHECKER_OK &&
            alternate_check == CHECKER_ERR_INVALID_CHECKSUM &&
            active_value == initial &&
            checkpoint_value == initial) {
            increment_passes();
        } else {
            increment_failures();
        }
        return;
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE_AND_CHECKPOINT_CHECKSUM:
        if (recovery_status == HARNESS_RECOVERY_RESTORE_FAILED &&
            checkpoint_check == CHECKER_OK &&
            primary_check == CHECKER_ERR_ABOVE_MAX &&
            restore_check == CHECKER_ERR_INVALID_CHECKSUM &&
            alternate_check == CHECKER_OK &&
            active_value == harness_fault_value) {
            increment_passes();
        } else {
            increment_failures();
        }
        return;
    default:
        if (recovery_status == HARNESS_RECOVERY_PRIMARY_ACCEPTED &&
            checkpoint_check == CHECKER_OK &&
            primary_check == CHECKER_OK &&
            restore_check == CHECKER_OK &&
            alternate_check == CHECKER_OK &&
            active_value == expected &&
            checkpoint_value == expected) {
            increment_passes();
        } else {
            increment_failures();
        }
        return;
    }
}

void harness_main(void) {
    harness_stage = HARNESS_STAGE_BOOT;

    for (;;) {
        const uint32_t iteration = harness_iteration + 1u;
        const uint32_t initial = sample_initial_value(iteration);
        const uint32_t expected = recovery_block_sample_primary_value(iteration);
        const checker_record_t record = sample_record(initial);
        recovery_block_sample_update_t update = {
            iteration,
            RECOVERY_BLOCK_SAMPLE_FAULT_NONE,
        };
        recovery_block_result_t result;

        harness_iteration = iteration;
        harness_last_initial_value = initial;
        harness_last_expected = expected;
        harness_last_value = 0u;
        harness_last_status = HARNESS_RECOVERY_PRIMARY_ACCEPTED;
        harness_last_recovery_status = HARNESS_RECOVERY_PRIMARY_ACCEPTED;
        harness_last_checkpoint_check = CHECKER_OK;
        harness_last_primary_check = CHECKER_OK;
        harness_last_restore_check = CHECKER_OK;
        harness_last_alternate_check = CHECKER_OK;
        harness_last_fault_target = HARNESS_FAULT_NONE;
        harness_c_recovery_block_state = checkpoint_record_init(record);
        mirror_state();

        harness_stage = HARNESS_STAGE_BEFORE_RECOVERY;
        harness_injection_point_before_recovery();

        harness_last_fault_target = harness_fault_target;
        result = recovery_block_run_with_hooks(
            (checkpoint_record_t *)&harness_c_recovery_block_state,
            recovery_block_sample_primary,
            apply_after_primary_fault,
            recovery_block_sample_alternate,
            apply_after_alternate_fault,
            &update);

        harness_last_recovery_status = result.status;
        harness_last_status = result.status;
        harness_last_checkpoint_check = result.checkpoint_check;
        harness_last_primary_check = result.primary_check;
        harness_last_restore_check = result.restore_check;
        harness_last_alternate_check = result.alternate_check;
        mirror_state();
        harness_last_value = harness_last_active_value;
        harness_fault_target = HARNESS_FAULT_NONE;

        harness_stage = HARNESS_STAGE_AFTER_RECOVERY;
        validate(initial, expected);
        harness_injection_point_after_recovery();
    }
}
